### ISSUE TITLE
[META] Implement printing and other remaining bidirectional capabilities

### ISSUE THREAD

**Issue by ehfd** [selkies-project/selkies#76]
Can be AI-driven: Yes, as long as all cases are tested carefully in real-life by people as well as automated tests.
Expected to be workable with: Fable 5, GPT-5.6 Sol, or equivalent (max thinking, requires expert guidance, especially with Opus 5, Kimi-K3, or under).

---

**Other types of bidirectional media:**

Printing: Implement a CUPS printing server > Convert received document to PDF/other adequate stateless format > Transport with HTTP API > Print from Web Browser

**Comment by ehfd** [comment]
This issue is from a request by @GeorgeDeac .

**Comment by ehfd** [comment]
Implemented support for Stereo on Opus. https://github.com/selkies-project/selkies-gstreamer/commit/1845bda05e0ed4baee9d5c6ac10ff884e2bd008f

**Comment by ehfd** [comment]
Good references:

https://github.com/evshiron/parsec-web-client
https://parsec.app/blog/game-streaming-tech-in-the-browser-with-parsec-5b70d0f359bc
https://github.com/parsec-cloud/parsec-sdk

**Comment by satyaakam** [comment]
Hi , i see this issue labelled under Funding , can anyone from core developers let me know what amount of funding are you expecting in order to see this issue resolved. Need few details like  amount , timelines , and process of resolving it in order to get the ball rolling .

**Comment by ehfd** [comment]
Hi! @satyaakam 

Unfortunately, the core developers are currently all busy, and no cost may justify the time we are to spend on this enhancement at this time.

However, @PMohanJ, another Indian developer and junior member of the Selkies team, is hired by his corporation to develop the webcam capabilities. I suggest waiting for them to open-source this capability.

**Comment by satyaakam** [comment]
Hi @ehfd thanks for the response , I am sorry if i sounded like someone who would like to throw money or people at a problem to get it resolved :-) , I understand the way FOSS communities work , my request was more from the point of view of collaboration, thanks for pointing to @PMohanJ  will be more than happy to collaborate and work towards getting it resolved .

**Comment by ehfd** [comment]
I greatly appreciate your interest in this project. @satyaakam 
I would have taken the job if I had extra time.
This project competes technically with well-funded proprietary remote desktop products and conversely relies on the free time of experienced developers in very specific, but wide-spectrum technology stacks. So donations never hurt.
But currently, we really need people willing to learn the code and make their own contributions.
Bounties might work out.

**Comment by debaa98** [comment]
is there any bounty going to be paid to contributer?

**Comment by ehfd** [comment]
@debaa98 It is possible, if you are indeed interested.

**Comment by debaa98** [comment]
Yaa I am intrested

**Comment by ehfd** [comment]
But I am not the entity which issues the bounty. Check with @satyaakam.

**Comment by debaa98** [comment]
yas,i had a contact with @satyaakam

**Comment by cardoza1991** [comment]
working on the meta stuff

**Comment by ehfd** [comment]
Please refer to the `features/websockets` branch and discuss with us in Discord. Microphone is implemented.
@cardoza1991

**Comment by ehfd** [comment]
Microphone and file upload/download capabilities are implemented. Webcam is still yet.

**Comment by debaa98** [comment]
hey @ehfd are you open for external contribution ?

**Comment by ehfd** [comment]
@debaa98 

Yes, but prior coordination in https://discord.gg/wDNGDeSW5F is highly recommended. Please tag @thelamer or @ehfd to get invited to a private contributor channel.

**Comment by debaa98** [comment]
@ehfd i guess fundng/Bounty is applicable for external contributers right ?

**Comment by ehfd** [comment]
@debaa98 
Yes, but it requires someone to pay the funding. We are not the providers of the bounty.

**Comment by to-a11y** [comment]
I built a Chromium-specific solution for this.

It uses CDP to intercept Page.fileChooserOpened, opens the local browser file picker, transfers the selected files to the remote container, and assigns them directly with DOM.setFileInputFiles.

Downloads work in the opposite direction using Chromium download events and are automatically forwarded to the local browser.

It supports multiple files, accept=, iframes and new tabs.

Repository:
[https://github.com/to-a11y/selkies-chromium-bridge](https://github.com/to-a11y/selkies-chromium-bridge?utm_source=chatgpt.com)

Demo:
https://github.com/user-attachments/assets/29881aa0-1274-4a53-a6b1-a3f6efed959a

This is Chromium-specific rather than a generic Selkies file-transfer mechanism, but from the user’s point of view upload/download behaves almost like a local browser.

**Comment by ehfd** [comment]
Implemented in https://github.com/selkies-project/selkies/commit/92efaf5cf187fb4f36a576c3502d3c68467ce8d1, https://github.com/selkies-project/selkies/commit/d637dda5790540a5276b26608203e9ce2be4396a, https://github.com/selkies-project/selkies/commit/862582500d3abab0e0334d951ecb9ee40cf9c5c5.
